cuda/test/components/searching.cu

```diff
@@ -0,0 +1,244 @@
+    std::shared_ptr<gko::CudaExecutor> cuda;
+    gko::Array<bool> result;
+    gko::Array<bool> dresult;
+    std::vector<int> sizes;
+};
+
+
+__device__ void test_assert(bool *success, bool predicate)
+{
+    if (!predicate) {
+        *success = false;
+    }
+}
+
+
+__global__ void test_binary_search(bool *success, int offset, int size)
+{
+    // test binary search on [0, size)
+    // for all possible partition points
+    auto result = binary_search(offset, size, [&](int i) {
+        // don't access out-of-bounds!
+        test_assert(success, i >= offset && i < offset + size);
+        return i >= threadIdx.x + offset;
+    });
+    auto result2 = binary_search(offset, size, [&](int i) {
+        // don't access out-of-bounds!
+        test_assert(success, i >= offset && i < offset + size);
+        return i >= threadIdx.x + offset + 1;
+    });
+    test_assert(success, result == threadIdx.x + offset);
+    test_assert(success, result2 == threadIdx.x + offset + 1);
+}
+
+TEST_F(Searching, BinaryNoOffset)
+{
+    run_test(test_binary_search, 0, config::warp_size);
+}
+
+TEST_F(Searching, BinaryOffset)
+{
+    run_test(test_binary_search, 5, config::warp_size);
+}
+
+
+__global__ void test_empty_binary_search(bool *success, int offset, int)
+{
+    auto result = binary_search(offset, 0, [&](int i) {
+        // don't access out-of-bounds!
+        test_assert(success, false);
+        return false;
+    });
+    test_assert(success, result == offset);
+}
+
+TEST_F(Searching, BinaryEmptyNoOffset)
+{
+    run_test(test_empty_binary_search, 0, 0);
+}
+
+TEST_F(Searching, BinaryEmptyOffset)
+{
+    run_test(test_empty_binary_search, 5, 0);
+}
+
+
+__global__ void test_sync_binary_search(bool *success, int, int size)
+{
+    // test binary search on [0, warp_size)
+    // for all possible partition points
+    auto result = synchronous_binary_search(size, [&](int i) {
+        // don't access out-of-bounds!
+        test_assert(success, i >= 0 && i < size);
+        return i >= threadIdx.x;
+    });
+    auto result2 = synchronous_binary_search(size, [&](int i) {
+        // don't access out-of-bounds!
+        test_assert(success, i >= 0 && i < size);
+        return i >= threadIdx.x + 1;
+    });
+    test_assert(success, result == threadIdx.x);
+    test_assert(success, result2 == threadIdx.x + 1);
+}
+
+TEST_F(Searching, SyncBinary)
+{
+    run_test(test_sync_binary_search, 0, config::warp_size);
+}
+
+
+__global__ void test_empty_sync_binary_search(bool *success, int, int)
+{
+    auto result = synchronous_binary_search(0, [&](int i) {
+        // don't access out-of-bounds!
+        test_assert(success, false);
+        return false;
+    });
+    test_assert(success, result == 0);
+}
+
+TEST_F(Searching, EmptySyncBinary)
+{
+    run_test(test_empty_sync_binary_search, 0, config::warp_size);
+}
+
+
+__global__ void test_warp_ary_search(bool *success, int offset, int size)
+{
+    // test binary search on [0, length)
```

Comment:
length is size I guess, also isn't it from offset rather than 0?

Reply:
Maybe these comments are better put in the respective TEST_F, since there is also a discrepancy  with the [0, warp_size) one.